core/src/main/java/jenkins/slaves/RemotingWorkDirSettings.java

```diff
@@ -0,0 +1,177 @@
+    
+    private final boolean disabled;
+    @CheckForNull
+    private final String  workDirPath;
+    @Nonnull
+    private final String  internalDir;
+    private final boolean failIfWorkDirIsMissing;
+
+    @DataBoundConstructor
+    public RemotingWorkDirSettings(boolean disabled, 
+            @CheckForNull String workDirPath, @CheckForNull String internalDir, 
+            boolean failIfWorkDirIsMissing) {
+        this.disabled = disabled;
+        this.workDirPath = Util.fixEmptyAndTrim(workDirPath);
+        this.failIfWorkDirIsMissing = failIfWorkDirIsMissing;
+        String internalDirName = Util.fixEmptyAndTrim(internalDir);
+        this.internalDir = internalDirName != null ? internalDirName : DEFAULT_INTERNAL_DIR;
+    }
+
+    public RemotingWorkDirSettings() {
+        // Enabled by default
+        this(false, null, DEFAULT_INTERNAL_DIR, false);
+    }
+
+    /**
+     * Check if workdir is disabled.
+     * 
+     * @return {@code true} if the property is disabled.
+     *         In such case Remoting will use the legacy mode.
+     */
+    public boolean isDisabled() {
+        return disabled;
+    }
+    
+    /**
+     * Indicates that agent root directory should be used as work directory.
+     * 
+     * @return {@code true} if the agent root should be a work directory.
+     */
+    public boolean isUseAgentRootDir() {
+        return workDirPath == null;
+    }
+
+    /**
+     * Check if startup should fail if the workdir is missing.
+     * 
+     * @return {@code true} if Remoting should fail if the the work directory is missing instead of creating it
+     */
+    public boolean isFailIfWorkDirIsMissing() {
+        return failIfWorkDirIsMissing;
+    }
+
+    /**
+     * Gets path to the custom workdir path.
+     * 
+     * @return Custom workdir path.
+     *         If {@code null}, an agent root directory path should be used instead.
+     */
+    @CheckForNull
+    public String getWorkDirPath() {
+        return workDirPath;
+    }
+
+    @Nonnull
+    public String getInternalDir() {
+        return internalDir;
+    }
+
+    @Override
+    public Descriptor<RemotingWorkDirSettings> getDescriptor() {
+        return Jenkins.getInstance().getDescriptor(RemotingWorkDirSettings.class);
+    }
+    
+    /**
+     * Gets a command line string, which can be passed to agent start command.
+     * 
+     * @param computer Computer, for which the arguments need to be constructed.
+     * @return Command line arguments. 
+     *         It may be empty if the working directory is disabled or
+     *         if the Computer type is not {@link SlaveComputer}.
+     */
+    @Nonnull
+    public String toCommandLineString(@Nonnull SlaveComputer computer) {
+        if(disabled) {
+            return "";
+        }
+        
+        StringBuilder bldr = new StringBuilder();
+        bldr.append("-workDir \"");
```

Comment:
Is this going to work with windows? I am not 100% sure but I feel like that backslash is for the start of a file in windows rather than a special character announcement